File: odoo_addons/flipkart_os/controllers/ai_assistant.py
# -*- coding: utf-8 -*-
from odoo import http
from odoo.http import request


class FlipkartAIAssistantController(http.Controller):

    @http.route('/flipkart_os/ai/bootstrap', type='json', auth='user')
    def bootstrap(self, session_id=None):
        Session = request.env['flipkart.ai.chat.session']
        sessions = Session.search([('user_id', '=', request.env.user.id)], limit=20)
        if session_id:
            current = Session.browse(int(session_id)).exists()
        else:
            current = sessions[:1]
        if not current:
            current = Session.create({'name': 'New AI Chat'})
            sessions = current | sessions
        return {
            'sessions': [self._session(session) for session in sessions],
            'current_session': self._session(current),
            'messages': [current._serialize_message(message) for message in current.message_ids.sorted('id')],
            'actions': [action._serialize() for action in current.proposed_action_ids],
            'settings': self._settings_status(),
        }

    @http.route('/flipkart_os/ai/session/new', type='json', auth='user')
    def new_session(self):
        session = request.env['flipkart.ai.chat.session'].create({'name': 'New AI Chat'})
        return {'session': self._session(session), 'messages': [], 'actions': []}

    @http.route('/flipkart_os/ai/session/load', type='json', auth='user')
    def load_session(self, session_id):
        session = request.env['flipkart.ai.chat.session'].browse(int(session_id)).exists()
        if not session:
            return {'error': 'Session not found.'}
        if session.user_id != request.env.user:
            return {'error': 'You cannot open this chat session.'}
        return {
            'session': self._session(session),
            'messages': [session._serialize_message(message) for message in session.message_ids.sorted('id')],
            'actions': [action._serialize() for action in session.proposed_action_ids],
        }

    @http.route('/flipkart_os/ai/session/delete', type='json', auth='user')
    def delete_session(self, session_id):
        Session = request.env['flipkart.ai.chat.session']
        session = Session.browse(int(session_id)).exists()
        if not session:
            return {'error': 'Session not found.'}
        if session.user_id != request.env.user:
            return {'error': 'You cannot delete this chat session.'}
        session.unlink()
        sessions = Session.search([('user_id', '=', request.env.user.id)], limit=20)
        current = sessions[:1]
        if not current:
            current = Session.create({'name': 'New AI Chat'})
            sessions = current | sessions
        return {
            'sessions': [self._session(item) for item in sessions],
            'current_session': self._session(current),
            'messages': [current._serialize_message(message) for message in current.message_ids.sorted('id')],
            'actions': [action._serialize() for action in current.proposed_action_ids],
        }

    @http.route('/flipkart_os/ai/message/send', type='json', auth='user')
    def send_message(self, session_id, content):
        session = request.env['flipkart.ai.chat.session'].browse(int(session_id)).exists()
        if not session:
            return {'error': 'Session not found.'}
        if session.user_id != request.env.user:
            return {'error': 'You cannot use this chat session.'}
        try:
            assistant_message = session.send_user_message(content)
        except Exception as error:
            return {
                'error': str(error),
                'session': self._session(session),
                'messages': [session._serialize_message(message) for message in session.message_ids.sorted('id')],
                'actions': [action._serialize() for action in session.proposed_action_ids],
            }
        return {
            'session': self._session(session),
            'messages': [session._serialize_message(message) for message in session.message_ids.sorted('id')],
            'assistant_message': assistant_message,
            'actions': [action._serialize() for action in session.proposed_action_ids],
        }

    @http.route('/flipkart_os/ai/action/approve', type='json', auth='user')
    def approve_action(self, action_id):
        action = request.env['flipkart.ai.proposed.action'].browse(int(action_id)).exists()
        if not action:
            return {'error': 'Action not found.'}
        try:
            return {'action': action.action_approve()}
        except Exception as error:
            return {'error': str(error), 'action': action._serialize()}

    @http.route('/flipkart_os/ai/action/reject', type='json', auth='user')
    def reject_action(self, action_id):
        action = request.env['flipkart.ai.proposed.action'].browse(int(action_id)).exists()
        if not action:
            return {'error': 'Action not found.'}
        return {'action': action.action_reject()}

    def _session(self, session):
        return {
            'id': session.id,
            'name': session.name,
            'write_date': session.write_date.isoformat() if session.write_date else None,
        }

    def _settings_status(self):
        settings = request.env['flipkart.ai.chat.session']._settings()
        return {
            'enabled': settings['enabled'],
            'provider_name': settings['provider_name'],
            'api_url_configured': bool(settings['api_url']),
            'model': settings['model'],
            'api_key_configured': bool(settings['api_key']),
            'api_format': settings['api_format'],
        }

File: odoo_addons/flipkart_os/static/src/js/ai_assistant.js
/** @odoo-module **/

import { Component, onPatched, onWillStart, useRef, useState } from "@odoo/owl";
import { registry } from "@web/core/registry";
import { rpc } from "@web/core/network/rpc";
import { useService } from "@web/core/utils/hooks";

class FlipkartAIAssistant extends Component {
    static template = "flipkart_os.AIAssistant";

    setup() {
        this.notification = useService("notification");
        this.action = useService("action");
        this.messagesRef = useRef("messages");
        this.state = useState({
            sessions: [],
            currentSession: null,
            messages: [],
            actions: [],
            settings: {},
            input: "",
            loading: false,
            error: "",
            sidebarOpen: window.innerWidth > 768,
        });
        onWillStart(async () => {
            await this.loadBootstrap();
        });
        onPatched(() => this.scrollMessagesToBottom());
    }

    async loadBootstrap(sessionId = null) {
        const data = await rpc("/flipkart_os/ai/bootstrap", { session_id: sessionId });
        this.state.sessions = data.sessions || [];
        this.state.currentSession = data.current_session || null;
        this.state.messages = data.messages || [];
        this.state.actions = data.actions || [];
        this.state.settings = data.settings || {};
        this.scrollMessagesToBottom();
    }

    async newSession() {
        const data = await rpc("/flipkart_os/ai/session/new", {});
        this.state.currentSession = data.session;
        this.state.messages = data.messages || [];
        this.state.actions = data.actions || [];
        await this.loadBootstrap(data.session.id);
    }

    async loadSession(session) {
        if (!session || this.state.loading) return;
        const data = await rpc("/flipkart_os/ai/session/load", { session_id: session.id });
        if (data.error) {
            this.notify(data.error, "danger");
            return;
        }
        this.state.currentSession = data.session;
        this.state.messages = data.messages || [];
        this.state.actions = data.actions || [];
        this.closeSidebarOnMobile();
    }

    async deleteCurrentSession() {
        if (!this.state.currentSession || this.state.loading) return;
        if (!window.confirm("Delete this chat history?")) return;
        const data = await rpc("/flipkart_os/ai/session/delete", { session_id: this.state.currentSession.id });
        if (data.error) {
            this.notify(data.error, "danger");
            return;
        }
        this.state.sessions = data.sessions || [];
        this.state.currentSession = data.current_session || null;
        this.state.messages = data.messages || [];
        this.state.actions = data.actions || [];
        this.notify("Chat deleted.", "success");
        this.closeSidebarOnMobile();
    }

    async sendMessage() {
        const content = (this.state.input || "").trim();
        if (!content || !this.state.currentSession || this.state.loading) return;
        this.state.loading = true;
        this.state.error = "";
        this.state.input = "";
        this.state.messages.push({ role: "user", content });
        try {
            const data = await rpc("/flipkart_os/ai/message/send", {
                session_id: this.state.currentSession.id,
                content,
            });
            if (data.error) {
                throw new Error(data.error);
            }
            this.state.currentSession = data.session || this.state.currentSession;
            this.state.messages = data.messages || this.state.messages;
            this.state.actions = data.actions || [];
            await this.loadBootstrap(this.state.currentSession.id);
            this.scrollMessagesToBottom();
        } catch (error) {
            this.state.error = error.message || String(error);
            this.notify(this.state.error, "danger");
        } finally {
            this.state.loading = false;
        }
    }

    async usePrompt(prompt) {
        this.state.input = prompt;
        await this.sendMessage();
    }

    async approveAction(item) {
        if (!item || item.state !== "proposed") return;
        try {
            const data = await rpc("/flipkart_os/ai/action/approve", { action_id: item.id });
            if (data.error) throw new Error(data.error);
            await this.refreshCurrent();
            this.notify("Draft action created.", "success");
        } catch (error) {
            this.notify(error.message || String(error), "danger");
            await this.refreshCurrent();
        }
    }

    async rejectAction(item) {
        if (!item || item.state !== "proposed") return;
        const data = await rpc("/flipkart_os/ai/action/reject", { action_id: item.id });
        if (data.error) {
            this.notify(data.error, "danger");
        }
        await this.refreshCurrent();
    }

    async refreshCurrent() {
        if (this.state.currentSession) {
            await this.loadSession(this.state.currentSession);
        }
    }

    openResult(item) {
        if (!item.result_model || !item.result_res_id) return;
        this.action.doAction({
            type: "ir.actions.act_window",
            res_model: item.result_model,
            res_id: item.result_res_id,
            views: [[false, "form"]],
            target: "current",
        });
    }

    notify(message, type = "info") {
        this.notification.add(message, { type });
    }

    onKeydown(ev) {
        if (ev.key === "Enter" && !ev.shiftKey) {
            ev.preventDefault();
            this.sendMessage();
        }
    }

    toggleSidebar() {
        this.state.sidebarOpen = !this.state.sidebarOpen;
    }

    closeSidebarOnMobile() {
        if (window.innerWidth <= 768) {
            this.state.sidebarOpen = false;
        }
    }

    scrollMessagesToBottom() {
        const element = this.messagesRef.el;
        if (!element) return;
        window.requestAnimationFrame(() => {
            element.scrollTop = element.scrollHeight;
        });
    }

    get pendingActions() {
        return (this.state.actions || []).filter((item) => item.state === "proposed");
    }

    get pastActions() {
        return (this.state.actions || []).filter((item) => item.state !== "proposed");
    }
}

registry.category("actions").add("flipkart_ai_assistant", FlipkartAIAssistant);
